etl/types/trasnformer_impl.go

```diff
@@ -0,0 +1,186 @@
+
+		hash, err := hash(records)
+		if err != nil {
+			return nil, errors.Wrapf(err, "cannot calculate JetDrop hash")
+		}
+		localJetDrop := &JetDrop{
+			Sections: sections,
+			RawData:  hash,
+		}
+		result = append(result, localJetDrop)
+	}
+
+	return result, nil
+}
+
+// hash calculate hash from record's hash
+func hash(r []Record) ([]byte, error) {
+	l := len(r)
+	data := make([][]byte, l)
+	for i, record := range r {
+		data[i] = record.Hash
+	}
+	return utils.Hash(data)
+}
+
+func getPulseData(rec *exporter.Record) (Pulse, error) {
+	r := rec.GetRecord()
+	pulse := r.ID.Pulse()
+	time, err := pulse.AsApproximateTime()
+	if err != nil {
+		return Pulse{}, errors.Wrapf(err, "could not get pulse ApproximateTime")
+	}
+	return Pulse{
+		PulseNo:        int(pulse.AsUint32()),
+		EpochPulseNo:   int(pulse.AsEpoch()),
+		PulseTimestamp: time,
+		NextPulseDelta: int(pulse.Next(pulseDelta)),
+		PrevPulseDelta: int(pulse.Prev(pulseDelta)),
+	}, nil
+}
+
+func getRecords(records []*exporter.Record) (map[insolar.JetID][]Record, error) {
+	// map need to collect records by JetID
+	res := make(map[insolar.JetID][]Record)
+	for _, r := range records {
+		record, err := transferToCanonicalRecord(r)
+		if err != nil {
+			return res, err
+		}
+		// collect records with some jetID
+		res[r.Record.JetID] = append(res[r.Record.JetID], record)
+	}
+
+	return res, nil
+}
+
+func transferToCanonicalRecord(r *exporter.Record) (Record, error) {
+	var (
+		recordType          RecordType
+		ref                 Reference
+		objectReference     Reference
+		prototypeReference  Reference = make([]byte, 0)
+		prevRecordReference Reference = make([]byte, 0)
+		recordPayload       []byte    = make([]byte, 0)
+		hash                []byte
+		rawData             []byte
+	)
+
+	ref = r.Record.ID.Bytes()
+	hash = r.Record.ID.Hash()
+	objectReference = r.Record.ObjectID.Hash()
+	dAtA, err := r.Marshal()
+	if err != nil {
+		return Record{}, errors.Wrapf(err, "cannot get record raw data")
+	}
+	rawData = dAtA
+
+	virtual := r.GetRecord().Virtual
+	switch virtual.Union.(type) {
+	case *ins_record.Virtual_Activate:
+		recordType = STATE
+		activate := virtual.GetActivate()
+		prototypeReference = activate.Image.Bytes()
+		recordPayload = activate.Memory
+		prevRecordReference = activate.PrevStateID().AsBytes()
+
+	case *ins_record.Virtual_Amend:
+		recordType = STATE
+		amend := virtual.GetAmend()
+		prototypeReference = amend.Image.Bytes()
+		recordPayload = amend.Memory
+		prevRecordReference = amend.PrevState.AsBytes()
+
+	case *ins_record.Virtual_Deactivate:
+		recordType = STATE
+		deactivate := virtual.GetDeactivate()
+		prototypeReference = deactivate.GetImage().AsBytes()
+		prevRecordReference = deactivate.PrevState.AsBytes()
+
+	case *ins_record.Virtual_Result:
+		recordType = RESULT
+		recordPayload = virtual.GetResult().Payload
+
+	case *ins_record.Virtual_IncomingRequest:
+		recordType = REQUEST
+		request := virtual.GetIncomingRequest()
```

Comment:
You can write it in one line as in next case:
`virtual.GetIncomingRequest().GetObject()`